async_simple/Signal.h

```diff
@@ -0,0 +1,453 @@
+/*
+ * Copyright (c) 2024, Alibaba Group Holding Limited;
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef ASYNC_SIMPLE_SIGNAL_H
+#define ASYNC_SIMPLE_SIGNAL_H
+
+#ifndef ASYNC_SIMPLE_USE_MODULES
+
+#include <assert.h>
+#include <any>
+#include <atomic>
+#include <bit>
+#include <cstdint>
+#include <memory>
+#include <system_error>
+#include <type_traits>
+#include <unordered_map>
+#include <utility>
+
+#include "async_simple/Common.h"
+#include "util/move_only_function.h"
+#endif  // ASYNC_SIMPLE_USE_MODULES
+
+namespace async_simple {
+
+enum SignalType : uint64_t {
+    none = 0,
+    // low 32 bit signal only trigger once.
+    // regist those signal's handler after triggered will failed and return
+    // false.
+    terminate = 1,  // signal type: terminate
+    // 1-16 bits reserve for async-simple
+    // 17-32 bits could used for user-defined signal
+
+    // high 32 bit signal, could trigger multiple times, emplace an handler
+    // 33-48 bits reserve for async-simple
+    // 49-64 bits could used for user-defined signal
+    // after signal triggered will always success and return true.
+    all = UINT64_MAX,
+};
+
+class Slot;
+
+class Signal : public std::enable_shared_from_this<Signal> {
+private:
+    using Handler = util::move_only_function<void(SignalType, Signal*)>;
+    using TransferHandler = util::move_only_function<void(SignalType)>;
+    struct HandlerManager {
+        uint8_t _index;
+        std::atomic<Handler*> _handler;
+        HandlerManager* _next;
+        HandlerManager(uint8_t index, HandlerManager* next)
+            : _index(index), _next(next) {}
+        void releaseNodes() {
+            for (HandlerManager* tmp = _next; tmp != nullptr; tmp = _next) {
+                _next = _next->_next;
+                delete tmp;
+            }
+            delete this;
+        }
+        ~HandlerManager();
+        static inline Handler emittedTag;
+    };
+    struct PrivateConstructTag {};
+    struct Node {
```

Comment:
`Node` is actually a shared state between `Slot` and `Signal`. It is not good to make it a subclass of `Signal`. Let's move it and renaming to `SignalSlotSharedState`. The list of `SignalSlotSharedState` can be another data structure.